explainer-datamodel.md

```diff
@@ -1,236 +1,329 @@
 
 ```js
-textrun2.styleMap.set("text-decoration", "underline");
+FormattedText.format( [ "The quick ", 
+                        { text: "brown", style: "color:brown;font-weight:bold" },
+                        " fox jumps over the lazy dog." 
+                      ], "font-style:italic" );
 ```
 
-The `styleMap` property is available on both the `FormattedText` and `FormattedTextRun` objects.
-
-It is also possible to set multiple styles at once similar to how HTML parses the `style` attribute.
-Set styles all at once at construction time:
+And in the following example, all the text will be blue _except_ for the word brown, which
+will be colored brown:
 
 ```js
-let textrun2 = new FormattedTextRun( { style: "color: yellow; font: 15pt Verdana" } );
+FormattedText.format( [ "The quick ", 
+                        { text: "brown", style: "color:brown" }, 
+                        " fox jumps over the lazy dog."
+                      ], { style: "color:blue" } );
 ```
 
-Or set groups of styles together dynamically afterward:
+A wide range of inline layout-related CSS is supported as style input to the `format` API.
 
-```js
-textrun2.setStyle( "color: yellow; font: 15pt Verdana" );
-```
+Styles specified in the second parameter apply to all text runs except where overridden on
+individual text objects (in the first parameter).
 
-### Cascading of values from FormattedText to text runs
+### Specifying constraints
 
-CSS properties applied to the style map from the FormattedText that are specified to inherit
-from parent to child will do so from the FormattedText to it's text run child objects.
+The final input to `format` are any constraints that should be applied to the formatted text.
+The primary constraints of interest to the API are the width and height of the containing block:
+how many CSS pixels are available for text layout.
 
-## Simple Example
+The layout flow will follow any writing-mode direction specified in the meta object, so in
+`horizontal-lr` (e.g., English) a width constraint on the containing block will cause line 
+wrapping in the inline direction; overflow will occur in the vertical direction. For specified 
+writing-modes such as `vertical-rl` (e.g., Chinese) a width constraint on the containing block 
+will only affect where overflow occurs--the line will not wrap because it is assumed to have 
+infinite vertical space to layout. To cause wrapping with `vertical-rl` writing modes, specify
+a height constraint instead.
 
-In this example, we create a `FormattedText` object and its text runs that we eventually 
-want to render with the text "The quick **brown** fox jumps over the lazy dog" where the
-word "brown" is colored brown and bold. In the [rendering explainer](explainer-rendering.md) we
-add layout constraints and render it to a canvas; at present, while it resides in the data model
-it has no associated layout.
+Constraints are specified on a constraint object using a `width` or `height` property:
```

Comment:
should we just allow an inline-size and disallow mixed writing modes? or is that too restrictive? In any case, inline-size and block-size might be better names here?

having a block-size for non-mixed writing modes seems to introduce questions about overflow.

Reply:
I think allowing for mixed-writing modes is potentially valuable (at least I don't want to artificially restrict those scenarios).

Re: the names `width` and `height`: I think of these values as physical area values that scope the layout engine's world--similar to constraints like the browser's window dimensions (screen width is very much independent of writing mode direction). Naturally these values _will_ constrain what layout can do in that space re: wrapping or overflow, and to me, it just depends on what writing-mode is applied at the context object level.

As an aside, we do need to think about overflow. There's an important scenario for some customer to be able to specify the block direction limit and `clip-path` to be able to clip content outside. We aught to think about what overflow means when specifying a restriction that interects with the writing-mode's block direction. In horizontal-lr, should layout continue to render content passed the "height" limit specified? If height can be compared to a screen with no scrollbar, than that is layout work that can never be seen; by not specifying a height, you are allowing for inifinite height. But specifying a height could imply clipping. Clipping becomes rather interesting in the metrics if we end up with partially rendered glyphs... ;)

In any case, I opened an issue to track this. #43